### PR TITLE
Support parsing for `allows` constraint

### PR DESCRIPTION
**Pull request by AlekseyTs**
https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md#ref-struct-generic-parameters

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @jjonescz, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs [hunk 1]
```diff
@@ -413,6 +413,8 @@ public enum SyntaxKind : ushort
         ScopedKeyword = 8448,
         /// <summary>Represents <see langword="file"/>.</summary>
         FileKeyword = 8449,
+        /// <summary>Represents <see langword="where"/>.</summary>
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:416]: > where

typo?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:416]: > > where
> 
> typo?

A copy/paste artifact. Will fix. Thanks.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs [hunk 2]
```diff
@@ -413,6 +413,8 @@ public enum SyntaxKind : ushort
...
+        AllowsKeyword = 8450,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:417]: nit: Consider adding the new syntax to `AllInOne` string or capture PROTOTYPE

**Review by jcouv — APPROVED** [review]
LGTM Thanks modulo typo already pointed out

File: src/Compilers/Test/Core/TestResource.resx [hunk 3]
```diff
@@ -554,7 +554,7 @@ namespace My
         [Obsolete]
         private volatile int f2;
         public abstract int m&lt;T&gt;(T t)
-          where T : struct
+          where T : struct, allows ref struct
```
Comment (AlekseyTs) [inline src/Compilers/Test/Core/TestResource.resx:557]:
```suggestion
          where T : struct
```

File: src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs [hunk 4]
```diff
@@ -0,0 +1,1648 @@
...
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void RefStruct_Single_MissingRefAndStruct()
+        {
+            var text = @"
+class C<T> where T : allows
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs:185]: Shouldn't this be an error?

File: src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs [hunk 5]
```diff
@@ -0,0 +1,1648 @@
...
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void RefStruct_Single_EscapedAllows()
+        {
+            var text = @"
+class C<T> where T : @allows ref struct
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs:232]: Consider testing also the other tokens escaped like `@ref` and `@struct`

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs:185]: > Shouldn't this be an error?

Why, it is not an error today.

File: src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs:232]: > Consider testing also the other tokens escaped like ```@ref``` and ```@struct```

Sure, I will add the sets in a follow up PR, unless I will have to make other changes in this PR

File: src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs (hunk 4 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Syntax/Parsing/AllowsConstraintParsing.cs:185]: oh, right, makes sense, thanks

**Pull request merged**